go/vt/vttablet/onlineddl/vrepl.go

```diff
@@ -0,0 +1,350 @@
+/*
+	Original copyright by GitHub as follows. Additions by the Vitess authors as follows.
+*/
+/*
+   Copyright 2016 GitHub Inc.
+	 See https://github.com/github/gh-ost/blob/master/LICENSE
+*/
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package onlineddl
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/dbconnpool"
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vttablet/onlineddl/vrepl"
+	"vitess.io/vitess/go/vt/vttablet/tabletmanager/vreplication"
+)
+
+// VReplStream represents a row in _vt.vreplication table
+type VReplStream struct {
+	id                   int64
+	workflow             string
+	source               string
+	pos                  string
+	timeUpdated          int64
+	transactionTimestamp int64
+	state                string
+	message              string
+	bls                  *binlogdatapb.BinlogSource
+}
+
+// VRepl is an online DDL helper for VReplication based migrations (ddl_strategy="online")
+type VRepl struct {
+	workflow    string
+	keyspace    string
+	shard       string
+	dbName      string
+	sourceTable string
+	targetTable string
+
+	sharedPKColumns *vrepl.ColumnList
+
+	sourceSharedColumns *vrepl.ColumnList
+	targetSharedColumns *vrepl.ColumnList
+	sharedColumnsMap    map[string]string
+
+	filterQuery string
+	bls         *binlogdatapb.BinlogSource
+
+	parser *vrepl.AlterTableParser
+}
+
+// NewVRepl creates a VReplication handler for Online DDL
+func NewVRepl(workflow, keyspace, shard, dbName, sourceTable, targetTable string) *VRepl {
+	return &VRepl{
+		workflow:    workflow,
+		keyspace:    keyspace,
+		shard:       shard,
+		dbName:      dbName,
+		sourceTable: sourceTable,
+		targetTable: targetTable,
+		parser:      vrepl.NewAlterTableParser(),
+	}
+}
+
+// getCandidateUniqueKeys investigates a table and returns the list of unique keys
+// candidate for chunking
+func (v *VRepl) getCandidateUniqueKeys(ctx context.Context, conn *dbconnpool.DBConnection, tableName string) (uniqueKeys [](*vrepl.UniqueKey), err error) {
```

Comment:
Is this function unused and work in progress? `sqlShowColumnsFrom` has only one parameter and four parameters are being bound. Also that query returns a different set of columns from the ones expected here.

Reply:
This function is not used right now. It is import from `gh-ost` as-is and is one of the things I will refactor later on.
However, there's a discussion we need to make about `vreplication` running on non-`PRIMARY` keys, something `gh-ost` supports and `vreplication` does not. So, I suspect this function may yet be used in the future.

Reply:
removed evaluation of transaction-timestamp